File: comparacao/prelude/bench.cpp
#include <chrono>
#include <ctime>
#include <iostream>
#include <vector>

#include "prelude.h"

#define benchmark(str, func)                   \
  do {                                         \
    auto start = std::clock();                 \
    (void) func;                               \
    auto stop = std::clock();                  \
    auto duration = 1000                       \
                  * (stop - start)             \
                  / CLOCKS_PER_SEC;            \
    std::cout << str << ": " << duration       \
              << " ms" << std::endl;           \
  } while (0)

int main (void)
{
    std::vector<int> input_vector;
    int n = 0;
    while (std::cin >> n)
        input_vector.push_back(n);

    auto dbl = [] (int x){ return x + x;};
    auto even = [](int x) { return x % 2 == 0; };

    benchmark("filter even",               filter(even, input_vector));
    benchmark("map (*2)",                  map(dbl, input_vector));
    benchmark("reverse",                   reverse(input_vector));
    benchmark("uncurry zip . split id id", zip(input_vector, input_vector));

    return 0;
}

File: comparacao/prelude/bench.hs
{-# LANGUAGE BangPatterns #-}

module Main where

import Control.DeepSeq
import Control.Monad.IO.Class (MonadIO(liftIO))
import System.CPUTime
import Text.Printf

split f g x = (f x, g x)
deepforce x = deepseq x x

readInput :: String -> [Int]
readInput = map (read :: String -> Int) . lines

-- Adapted from https://github.com/stackbuilders/reverse/blob/fa120ba688aea328a587cbb1971280c22085a257/benchmarks/ClockBenchmarks.hs
timeSomething :: NFData a => String -> a -> IO ()
timeSomething str something = do
  start <- liftIO getCPUTime
  let !result = deepforce $! something
  end <- liftIO getCPUTime
  let diff = round . (/1000000000) . fromIntegral $ end - start
  putStrLn $ str ++ ": " ++ show diff ++ " ms"

main :: IO ()
main = do
    input <- getContents
    let !l = deepforce $ readInput input

    timeSomething "filter even"               $ filter even l
    timeSomething "map (*2)"                  $ map (*2) l
    timeSomething "reverse"                   $ reverse l
    timeSomething "uncurry zip . split id id" $ zip l l
